crates/lib/src/qpu/experimental/random.rs

```diff
@@ -0,0 +1,272 @@
+                "source".to_string(),
+                false,
+                ExternParameterType::VariableLengthVector(quil_rs::instruction::ScalarType::Real),
+            ),
+            ExternParameter::try_new(
+                "sub_region_size".to_string(),
+                false,
+                ExternParameterType::Scalar(quil_rs::instruction::ScalarType::Integer),
+            ),
+            ExternParameter::try_new(
+                "seed".to_string(),
+                true,
+                ExternParameterType::Scalar(quil_rs::instruction::ScalarType::Integer),
+            ),
+        ]
+        .into_iter()
+        .map(|r| r.map_err(Error::from))
+        .collect::<Result<Vec<ExternParameter>>>()?;
+        Ok(quil_rs::instruction::ExternSignature::new(None, parameters))
+    }
+}
+
+/// Hardware values are 48 bits long.
+const MAX_SEQUENCER_VALUE: u64 = 0xFFFF_FFFF_FFFF;
+
+/// Hardware multiplication currently uses the lower 16 bits of
+/// the PRNG value.
+const MAX_UNSIGNED_MULTIPLIER: u64 = 0x0000_0000_FFFF;
+
+/// A valid seed value that may be used to initialize the PRNG. Such
+/// values are in the range `[1, MAX_SEQUENCER_VALUE]` and are losslessly
+/// convertible to `f64`.
+#[derive(Debug, Clone, Copy)]
+pub struct PrngSeedValue {
+    as_u64: u64,
+    pub(super) as_f64: f64,
+}
+
+impl PrngSeedValue {
+    /// Attempt to create a new instance of `PrngSeedValue` from a `u64`.
+    /// The value must be in the range `[1, MAX_SEQUENCER_VALUE]` and
+    /// losslessly convertible to `f64`.
+    pub fn try_new(value: u64) -> Result<Self> {
+        if !(1..=MAX_SEQUENCER_VALUE).contains(&value) {
+            return Err(Error::InvalidSeed(value));
+        }
+        if let Some(f64_value) = value.to_f64() {
+            Ok(Self {
+                as_u64: value,
+                as_f64: f64_value,
+            })
+        } else {
+            Err(Error::InvalidSeed(value))
+        }
+    }
+}
+
+fn lfsr_next(seed: u64, taps: &[u32]) -> u64 {
+    let feedback_value = taps.iter().fold(0, |acc, tap| {
+        let base = 2u64.pow(*tap - 1);
+        let bit = u64::from((seed & base) != 0);
+        acc.bitxor(bit)
+    });
+    (seed << 1) & MAX_SEQUENCER_VALUE | feedback_value
+}
+
+/// This represents the LFSR currently implemented on Rigetti control systems. Specifically,
+/// it implements a 48-bit LFSR with taps at indices 48, 47, 21, and 20.
+#[must_use]
+pub fn lfsr_v1_next(seed: PrngSeedValue) -> u64 {
+    lfsr_next(seed.as_u64, &[48, 47, 21, 20])
+}
+
+fn generate_lfsr_v1_sequence(seed: u64, start_index: u32, series_length: u32) -> Vec<u64> {
+    let mut lfsr = seed & MAX_SEQUENCER_VALUE;
+
+    let range = start_index..(start_index + series_length);
+    let mut collection = vec![];
+    for i in 0..(start_index + series_length) {
+        lfsr = lfsr_next(lfsr, &[48, 47, 21, 20]);
+        if range.contains(&i) {
+            collection.push(lfsr);
+        }
+    }
+    collection
+}
```

Comment:
Looking forward a bit:

It looks like we expect the possibility of further LFSR revisions, since we have a `v1` naming scheme here. Would it be worth refactoring this code to make it easier to add a `v2`? Thinking something like

```rust
trait LfsrVersion {
    const TAPS: &[u32];
}

fn lfsr_next<LFSR: LfsrVersion>(seed: PrngSeedValue) -> u64 {
    // Same as implemented here, but replace `taps` with `LFSR::TAPS`.
}

fn generate_lfsr_sequence<LFSR: LfsrVersion>(seed: u64, start_index: u32, series_length: u32) -> Vec<u64> {
    // Same as implemented here, but call `lfsr_next::<LFSR>(lfsr)` instead.
}
```

(Alternatively, use an `enum` with an associated `taps(&self)` method, but the idea is the same)

Reply:
I've implemented this locally, but I actually think the simple functions are better in this case; including the trait in the signature requires that the trait is public, and there's no real reason to support public implementations at this point.

Reply:
I'm fine leaving this as-is, especially if we aren't expecting a v2 any time soon.

For future reference, we wouldn't have to worry about public implementations if we made `LfsrVersion` a _sealed_ trait (publicly visible, not publicly implementable)